### PR TITLE
Add option to report empty GC disc drive as closed (issue 11840)

### PR DESCRIPTION
**Pull request by blaahaj**
https://bugs.dolphin-emu.org/issues/11840

This gives important doot-doot-doot music and bouncy cube.

Only problem: when an empty disc drive is reported as closed, the IPL encounters a read error:

> The disc could not be read.
>
> Please read the NINTENDO GAMECUBE  
> Instruction Booklet for more information.

This means the menu does not load afterwards.

I don't know why that happens. I wasn't able to fix it. So setting is off by default.

### CODE REVIEW

**Comment by mbc07** [comment]
Unless I misread the issue report, having an IPL dump, disabling "Skip Main Menu" option in Dolphin's settings then holding B button on GC Controller 1 when booting any GameCube game should achieve the same, no?

**Comment by Techjar** [comment]
Yeah you misread the issue report. This applies when booting the IPL directly without a disc (Tools > Load GameCube Main Menu).

**Comment by blaahaj** [comment]
Do you think “The disc could not be read.” prevents merging this? I do not expect that this error is a regression. I think it is likelier that it was simply hidden before, because Dolphin did not ever report the drive as empty and closed at the same time when the IPL runs. The new option I add does not default to on, in order that it would not break anything for users unless they should choose to edit the setting.

**Comment by Pokechu22** [comment]
It's probably because the drive isn't returning the proper error code.  If I recall correctly from my earlier tests, the high byte should be 0x03000000 (`ERROR_NO_DISK_L`) and the low bytes should be 0x023a00 (`ERROR_NO_DISK_H`).  I think, at least.

Note that my tests were on the Wii, and were based on a rather silly edge case; the Wii _will_ report the cover being closed with no disc if you insert something other than a DVD in it, such as an audio CD.  This also causes the disc channel to be confused, same as in dolphin.  I don't have a gamecube (and as such, I don't have an IPL dump), so I can't say whether the same thing would work here.

You probably need to add another case here:

https://github.com/dolphin-emu/dolphin/blob/b29ba7119a6695f6302da4d0f59ea50d02ee65b7/Source/Core/Core/HW/DVD/DVDInterface.cpp#L718-L734

**Comment by blaahaj** [comment]
> It's probably because the drive isn't returning the proper error code. If I recall correctly from my earlier tests, the high byte should be 0x03000000 (`ERROR_NO_DISK_L`) and the low bytes should be 0x023a00 (`ERROR_NO_DISK_H`). I think, at least.

If I am remembering right, I did exactly what you said and it did not work. But, I will try this again, it could have been another combination of bits.

**Comment by Pokechu22** [comment]
Oh, I should also note that it should probably be returning `ERROR_COVER` for the high bits in the current case, and that I locally renamed `ERROR_COVER_H` to `ERROR_NO_DISK_H`, which is probably why you didn't run into that...

**Comment by blaahaj** [comment]
@Pokechu22 The current code `s_error_code = ERROR_NO_DISK | ERROR_COVER_H;` is already those hex values.

**Comment by Pokechu22** [comment]
Can you run it with the `DVD Interface (DVD)` log enabled, and then post the log file?  Also, can you post a screenshot of the disc could not be read screen?

**Comment by blaahaj** [comment]
@Pokechu22 

```
34:40:891 Core/HW/DVD/DVDInterface.cpp:928 I[DVD]: Read DiscID 00000000
34:40:909 Core/HW/DVD/DVDInterface.cpp:1070 I[DVD]: DVDLowStopMotor  kill!
```

Screenshot


I apologise that I did not see your comment sooner.

**Comment by Pokechu22** [comment]
I'm not entirely sure what to make of that, unfortunately.  It doesn't seem to have even requested the error, which doesn't match up with my expectations.

What happens if you boot the IPL with a Wii disc set as the default disc (both before and after this change)?

**Comment by blaahaj** [comment]
> What happens if you boot the IPL with a Wii disc set as the default disc (both before and after this change)?

* With 'Report disc drive as closed when empty' - same error message
* Without 'Report disc drive as closed when empty' - same lack of doot doot doot

Apologise again for slow response.

**Comment by blaahaj** [comment]
Rebased code.

It seems the error message was fixed?

**Comment by blaahaj** [comment]
Now:

* With 'Report disc drive as closed when empty' - doot doot doot. No error message. ✔️ 
* Without 'Report disc drive as closed when empty' - no doot doot doot. No error message. ✔️

PAL & NTSC-U IPL tested.

**Comment by blaahaj** [comment]
Could some one remove the " the WIP / do not merge "?

File: Source/Core/Core/HW/DVD/DVDInterface.cpp [hunk 1]
```diff
@@ -347,14 +347,23 @@ static void DTKStreamingCallback(DIInterruptType interrupt_type, const std::vect
   }
 }
 
+static bool ShouldLidBeOpen()
+{
+  // Disc Channel relies on cover being open when no disc is inserted
+  // The Wii also has no physical cover
```
Comment (MayImilae) [inline Source/Core/Core/HW/DVD/DVDInterface.cpp:353]: About that... 

wiimini

Of course we don't target the Wii Mini, so this is all semantics. Might be nice to include a nod to this in the comment though!

Anyway, I'll take a look at the UI for this shortly and do a follow up.

**Comment by Rumi-Larry** [comment]
What happened with the pr? :(

**Comment by Zopolis4** [comment]
What's the status of this?

**Comment by AdmiralCurtiss** [comment]
Can you rebase this?

**Comment by AdmiralCurtiss** [comment]
Rebased as #11570
